mm2src/trading_api/src/one_inch_api/types.rs

```diff
@@ -0,0 +1,416 @@
+        self.disable_estimate = disable_estimate;
+        self
+    }
+    pub fn with_allow_partial_fill(mut self, allow_partial_fill: Option<bool>) -> Self {
+        self.allow_partial_fill = allow_partial_fill;
+        self
+    }
+
+    pub fn build_query_params(&self) -> MmResult<QueryParams<'static>, ApiClientError> {
+        self.validate_params()?;
+
+        let mut params = vec![
+            ("src", self.src.clone()),
+            ("dst", self.dst.clone()),
+            ("amount", self.amount.clone()),
+            ("from", self.from.clone()),
+            ("slippage", self.slippage.to_string()),
+        ];
+
+        push_if_some!(params, "fee", self.fee);
+        push_if_some!(params, "protocols", &self.protocols);
+        push_if_some!(params, "gasPrice", &self.gas_price);
+        push_if_some!(params, "complexityLevel", self.complexity_level);
+        push_if_some!(params, "parts", self.parts);
+        push_if_some!(params, "mainRouteParts", self.main_route_parts);
+        push_if_some!(params, "gasLimit", self.gas_limit);
+        push_if_some!(params, "includeTokensInfo", self.include_tokens_info);
+        push_if_some!(params, "includeProtocols", self.include_protocols);
+        push_if_some!(params, "includeGas", self.include_gas);
+        push_if_some!(params, "connectorTokens", &self.connector_tokens);
+        push_if_some!(params, "permit", &self.permit);
+        push_if_some!(params, "receiver", &self.receiver);
+        push_if_some!(params, "referrer", &self.referrer);
+        push_if_some!(params, "disableEstimate", self.disable_estimate);
+        push_if_some!(params, "allowPartialFill", self.allow_partial_fill);
+
+        Ok(params)
+    }
+
+    /// Validate params by 1inch rules (to avoid extra requests)
+    fn validate_params(&self) -> MmResult<(), ApiClientError> {
+        validate_slippage(self.slippage)?;
+        validate_fee(&self.fee)?;
+        validate_complexity_level(&self.complexity_level)?;
+        validate_gas_limit(&self.gas_limit)?;
+        validate_parts(&self.parts)?;
+        validate_main_route_parts(&self.main_route_parts)?;
+        Ok(())
+    }
+}
+
+#[derive(Deserialize, Debug)]
+pub struct TokenInfo {
+    pub address: Address,
+    pub symbol: String,
+    pub name: String,
+    pub decimals: u32,
+    pub eip2612: bool,
+    #[serde(rename = "isFoT")]
+    pub is_fot: bool,
+    #[serde(rename = "logoURI")]
+    pub logo_uri: String,
+    pub tags: Vec<String>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProtocolInfo {
+    pub name: String,
+    pub part: f64,
+
+    #[serde(rename = "fromTokenAddress")]
+    pub from_token_address: Address,
+
+    #[serde(rename = "toTokenAddress")]
+    pub to_token_address: Address,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct ClassicSwapData {
+    /// dst token amount to receive, in api is a decimal number as string
+    #[serde(rename = "dstAmount")]
+    pub dst_amount: String,
```

Comment:
My intent was to have this lib to use the types exactly which 1inch requires. 
So to do all conversion at the upper rpc level (in ext_api module), to avoid duplicate conversion and error processing

Reply:
I see, you refactored it to this. Ok now we can be user that `1inch` api returns eth addresses

```rust
#[derive(Deserialize, Debug, Serialize)]
pub struct TokenInfo {
    pub address: Address,
    pub symbol: String,
    pub name: String,
    pub decimals: u32,
    pub eip2612: bool,
    #[serde(rename = "isFoT")]
    pub is_fot: bool,
    #[serde(rename = "logoURI")]
    pub logo_uri: String,
    pub tags: Vec<String>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ProtocolInfo {
    pub name: String,
    pub part: f64,
    #[serde(rename = "fromTokenAddress")]
    pub from_token_address: Address,
    #[serde(rename = "toTokenAddress")]
    pub to_token_address: Address,
}

#[derive(Deserialize, Debug)]
pub struct ClassicSwapData {
    /// dst token amount to receive, in api is a decimal number as string
    #[serde(rename = "dstAmount")]
    pub dst_amount: String,
    #[serde(rename = "srcToken")]
    pub src_token: Option<TokenInfo>,
    #[serde(rename = "dstToken")]
    pub dst_token: Option<TokenInfo>,
    pub protocols: Option<Vec<Vec<Vec<ProtocolInfo>>>>,
    pub tx: Option<TxFields>,
    pub gas: Option<u128>,
}